drivers/sensor/paw3212/paw3212.c

```diff
@@ -0,0 +1,602 @@
+#define T_SRX		(20 - T_NCS_SCLK)	/* 20 us */
+#define T_SCLK_NCS_WR	(35 - T_NCS_SCLK)	/* 35 us */
+#define T_SWX		(180 - T_SCLK_NCS_WR)	/* 180 us */
+#define T_SRAD		160			/* 160 us */
+
+
+/* Sensor registers */
+#define PAW3212_REG_PRODUCT_ID		0x00
+#define PAW3212_REG_REVISION_ID		0x01
+#define PAW3212_REG_MOTION		0x02
+#define PAW3212_REG_DELTA_X_LOW		0x03
+#define PAW3212_REG_DELTA_Y_LOW		0x04
+#define PAW3212_REG_POWER_UP_RESET	0x06
+#define PAW3212_REG_WRITE_PROTECT	0x09
+#define PAW3212_REG_DELTA_XY_HIGH	0x12
+#define PAW3212_REG_MOUSE_OPTION	0x19
+
+/* Motion status bits */
+#define PAW3212_MOTION_STATUS_MOTION	BIT(7)
+#define PAW3212_MOTION_STATUS_DYOVF	BIT(4)
+#define PAW3212_MOTION_STATUS_DXOVF	BIT(3)
+
+/* Convert deltas to x and y */
+#define PAW3212_DELTA_X(xy_high, x_low)	expand_s12((((xy_high) & 0xF0) << 4) | (x_low))
+#define PAW3212_DELTA_Y(xy_high, y_low)	expand_s12((((xy_high) & 0x0F) << 8) | (y_low))
+
+/* Sensor identification values */
+#define PAW3212_PRODUCT_ID		0x30
+
+/* Write protect magic */
+#define PAW3212_WPMAGIC			0x5A
+
+#define SPI_WRITE_BIT			BIT(7)
+
+
+enum async_init_step {
+	ASYNC_INIT_STEP_POWER_UP,
+	ASYNC_INIT_STEP_VERIFY_ID,
+	ASYNC_INIT_STEP_CONFIGURE,
+
+	ASYNC_INIT_STEP_COUNT
+};
+
+struct paw3212_data {
+	struct device                *cs_gpio_dev;
+	struct device                *irq_gpio_dev;
+	struct device                *spi_dev;
+	struct gpio_callback         irq_gpio_cb;
+	struct k_spinlock            lock;
+	s16_t                        x;
+	s16_t                        y;
+	sensor_trigger_handler_t     data_ready_handler;
+	struct k_delayed_work        init_work;
+	enum async_init_step         async_init_step;
+	int                          err;
+	bool                         ready;
+};
+
+static const struct spi_config spi_cfg = {
+	.operation = SPI_WORD_SET(8) | SPI_TRANSFER_MSB |
+		     SPI_MODE_CPOL | SPI_MODE_CPHA,
+	.frequency = CONFIG_DESKTOP_SPI_FREQ_HZ,
+	.slave = 0,
+};
+
+static const s32_t async_init_delay[ASYNC_INIT_STEP_COUNT] = {
+	[ASYNC_INIT_STEP_POWER_UP]  = 1,
+	[ASYNC_INIT_STEP_VERIFY_ID] = 0,
+	[ASYNC_INIT_STEP_CONFIGURE] = 0,
+};
+
+
+static int paw3212_async_init_power_up(struct paw3212_data *dev_data);
+static int paw3212_async_init_verify_id(struct paw3212_data *dev_data);
+static int paw3212_async_init_configure(struct paw3212_data *dev_data);
+
+
+static int (* const async_init_fn[ASYNC_INIT_STEP_COUNT])(struct paw3212_data *dev_data) = {
+	[ASYNC_INIT_STEP_POWER_UP]  = paw3212_async_init_power_up,
+	[ASYNC_INIT_STEP_VERIFY_ID] = paw3212_async_init_verify_id,
+	[ASYNC_INIT_STEP_CONFIGURE] = paw3212_async_init_configure,
+};
+
+
+static struct paw3212_data paw3212_data;
+DEVICE_DECLARE(paw3212);
+
+
+static s16_t expand_s12(s16_t x)
+{
+	/* Integer promotion will convert s16_t to int. To expand sign we need
+	 * to shift left and then right. Right shift will be arithmetic since
+	 * type is signed.
+	 */
+	return (x << 20) >> 20;
```

Comment:
It it well defined behaviour ?

Reply:
I rechecked and indeed this needs fixing as left shift of negative number is undefined. I expect nothing should really happen as compiler cannot assume that value is negative (and so cannot do any optimizations here) but it must be fixed.

Right shift of negative is implementation defined and GCC docs mention that sign is extended (exactly what we want to have here). As far as I checked other compilers do the same. I will try to place static assert to catch a possible problem with it.
btw - we have similar problem with implementation defined representation of negative values as we assume it will be 2s complement (i.e. we pass what sensor gives us without any conversion). I could assert that too just in case.

Reply:
Fixed. I have also added an assert that should check both this and negative numbers format.